Package updater as RPM & OCI

chouquette opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

This PR adds support for packaging the remote updater as RPM & OCI packages

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

Having the remote updater self-update, and allow initial install on RHEL derivatives

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

This is based on https://github.com/DataDog/datadog-agent/pull/24131, which should be merged first

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

chouquette committed:
packaging: add rules to package updater as RPM

chouquette committed:
add remote updater rpm deploy rule

chouquette committed:
fixup rpm packaging

chouquette committed:
update JOBOWNERS

chouquette committed:
packaging: prepare a template for OCI jobs

chouquette committed:
make OCI template product agnostic

chouquette committed:
package remote updater as OCI

chouquette committed:
deploy updater OCI

chouquette committed:
provide the same RPM settings as the agent

chouquette committed:
updater: provide same compression settings as other packages

chouquette committed:
deploy RPM updater for kitchen/e2e tests

chouquette committed:
fix package scripts packaging

chouquette committed:
updater: add RPM package script

raphaelgavache commented on omnibus/package-scripts/updater-rpm/postrm:
```diff
@@ -0,0 +1,34 @@
+echo "Force-deleting $INSTALL_DIR"
+rm -rf $INSTALL_DIR
+rm -rf $LOG_DIR
+rm -rf $CONFIG_DIR
```
the CONFIG_DIR on agent rpm uninstall is kept: https://github.com/DataDog/datadog-agent/blob/main/omnibus/package-scripts/agent-rpm/postrm

Should rpm remove be equivalent to apt purge?

chouquette replied:
I think it's up to the package authors to chose a behavior for config files.

The safe approach is probably to keep them for now

raphaelgavache reviewed (approved):
lgtm, I'll complete with extra e2e tests on top

chouquette committed:
updater: rpm: keep configuration files after uninstall

CelianR commented on .gitlab/deploy_packages/oci.yml:
```diff
@@ -0,0 +1,39 @@
+  extends: ".deploy_packages_oci"
+  needs: [ "agent_oci" ]
+  rules:
+    !reference [.on_deploy_a7]
```
```suggestion
    !reference [.on_deploy]
```

CelianR commented on .gitlab/deploy_packages/oci.yml:
```diff
@@ -0,0 +1,39 @@
+  extends: ".deploy_packages_oci"
+  needs: [ "updater_oci" ]
+  rules:
+    !reference [.on_deploy_a7]
```
```suggestion
    !reference [.on_deploy]
```

CelianR commented on .gitlab/kitchen_deploy/kitchen_deploy.yml:
```diff
@@ -245,6 +245,40 @@ deploy_rpm_testing-a7_arm64:
+deploy_rpm_testing-u7_arm64:
+  rules:
+    - !reference [.except_no_tests_no_deploy]
+    - !reference [.on_a7]
```
```suggestion
    - when: on_success
```

CelianR commented on .gitlab/kitchen_deploy/kitchen_deploy.yml:
```diff
@@ -245,6 +245,40 @@ deploy_rpm_testing-a7_arm64:
+deploy_rpm_testing-u7_x64:
+  rules:
+    - !reference [.except_no_tests_no_deploy]
+    - !reference [.on_a7]
```
```suggestion
    - when: on_success
```

CelianR commented on .gitlab/packaging/oci.yml:
```diff
@@ -28,39 +21,59 @@ packaging_oci:
+
+updater_oci:
+  extends: .package_oci
+  rules: !reference [.on_a7]
```
```suggestion
```

CelianR reviewed (changes requested):
Agent 7 rules have been removed, see this [PR](https://github.com/DataDog/datadog-agent/pull/24209) and related PRs

chouetz commented on .gitlab/kitchen_deploy/kitchen_deploy.yml:
```diff
@@ -245,6 +245,36 @@ deploy_rpm_testing-a7_arm64:
+  image: 486234852809.dkr.ecr.us-east-1.amazonaws.com/ci/datadog-agent-builders/gitlab_agent_deploy:$DATADOG_AGENT_BUILDERS
+  tags: ["arch:amd64"]
+  rules:
+    - !reference [.except_no_tests_no_deploy]
```
```suggestion
    - !reference [.except_no_tests_no_deploy]
    - !reference [.except_mergequeue]
    - when: on_success
```
The `except_mergeque` is probably not necessary but it would align all rules together. I'll do some cleanup later

chouetz commented on .gitlab/packaging/oci.yml:
```diff
@@ -1,25 +1,18 @@
-    KUBERNETES_MEMORY_LIMIT: "32Gi"
+  before_script:
+    - source /root/.bashrc
+    - export INSTALL_DIR=/opt/datadog-packages/datadog-agent/$(inv agent.version -u)-1
```
```suggestion
    - export INSTALL_DIR=/opt/datadog-packages/datadog-agent/$(inv agent.version -u --no-include-git --no-include-pre)-1
```
Not sure about this though..
Bonus: you can maybe also `export PACKAGE_VERSION` on `before_script` and use it in the definition of `INSTALL_DIR` if it's the one you really need, wdyt?

chouetz replied:
Plus, should it be `export INSTALL_DIR=/opt/datadog-packages/${OCI_PRODUCT}/$(inv agent.version -u)-1` instead?

chouquette replied:
Having `--no-include-git` and `--no-include-pre` is likely to cause conflicts for dev builds as they will all share the same version no?
You are 100% correct regarding the `${OCI_PRODUCT}` part, I'll fix this right away, and will take a look at your suggestion for `PACKAGE_VERSION`

chouquette replied:
Done!

chouetz commented on .gitlab/packaging/rpm.yml:
```diff
@@ -0,0 +1,47 @@
+
+.updater_package_rpm_common:
+  rules:
+    - !reference [.except_mergequeue]
```
```suggestion
    - !reference [.except_mergequeue]
    - when: on_success
```

chouetz commented on omnibus/package-scripts/updater-rpm/posttrans:
```diff
@@ -0,0 +1,75 @@
+#!/bin/sh
+
+readonly INSTALL_DIR=/opt/datadog/updater
```
seems `OCI_PRODUCT` is `datadog-updater`, could it be an issue here?

raphaelgavache replied:
the OCI package path and rpm/deb are currently different ones, /opt/datadog/updater is the right one for rpm

chouetz commented on omnibus/package-scripts/updater-rpm/posttrans:
```diff
@@ -0,0 +1,75 @@
+#!/bin/sh
+
+readonly INSTALL_DIR=/opt/datadog/updater
+readonly UPDATER_HELPER=${INSTALL_DIR}/bin/updater/updater-helper
```
Same as previous comment

Pythyu commented on omnibus/package-scripts/updater-rpm/posttrans:
```diff
@@ -0,0 +1,75 @@
+$INSTALL_DIR/bin/updater/updater bootstrap -P datadog-agent
+
+# start updater
+SYSTEMCTL_SKIP_SYSV=true systemctl enable datadog-updater || true
+SYSTEMCTL_SKIP_SYSV=true systemctl start datadog-updater || true
```
We probably should check if the local system is using systemctl or initctl or service like in [here](https://github.com/DataDog/datadog-agent/blob/74d25e20c1e886b78099ff45c612e2aba09aec4c/omnibus/package-scripts/agent/posttrans#L110-L119)

chouquette replied:
I think the updater doesn't intend to support anything else than systemd, as the experiment system it uses is heavily based on systemd units
I'll let @raphaelgavache confirm

raphaelgavache replied:
yes we only support systemd for the moment (and probably forever given that all major distros have moved to it)

Pythyu commented on omnibus/package-scripts/updater-rpm/preinst:
```diff
@@ -0,0 +1,5 @@
+#!/bin/sh
+
+SYSTEMCTL_SKIP_SYSV=true systemctl stop datadog-updater || true
```
same as earlier comment on systemctl

Pythyu commented on omnibus/package-scripts/updater-rpm/prerm:
```diff
@@ -0,0 +1,23 @@
+
+readonly INSTALL_DIR=/opt/datadog/updater
+
+SYSTEMCTL_SKIP_SYSV=true systemctl stop datadog-updater || true
+SYSTEMCTL_SKIP_SYSV=true systemctl disable datadog-updater || true
```
same as earlier comment on systemctl

Pythyu commented on omnibus/config/projects/updater.rb:
```diff
@@ -17,6 +17,17 @@
+else
+  COMPRESSION_THREADS = 1
+end
+if ENV.has_key?("DEPLOY_AGENT") && ENV["DEPLOY_AGENT"] == "true"
+  COMPRESSION_LEVEL = 9
+else
+  COMPRESSION_LEVEL = 5
+end
```
❓ Question :
Is it so that dev builds are faster than release ones ?

chouquette replied:
Yes, this is what we do for other projects.
The basic reasonning is that for deploy builds (ideally this should be based on a "is_release" flag but I don't think we have such a flag) we will have to pay for the bandwidth when users download the product, so we're fine spending a bit of extra time compressing it.
For internal builds, we don't really care about bandwidth and mostly want to have a build ready fast

Pythyu reviewed (commented):
Other than my comments looks good 👍

chouetz reviewed (commented):
Few updates on rules that are required I guess

chouquette committed:
Update .gitlab/kitchen_deploy/kitchen_deploy.yml

Co-authored-by: Nicolas Schweitzer <nicolas.schweitzer@datadoghq.com>

chouquette committed:
packaging: oci: don't hardcode product and refactor version fetching

chouquette committed:
Update .gitlab/packaging/rpm.yml

Co-authored-by: Nicolas Schweitzer <nicolas.schweitzer@datadoghq.com>

CelianR commented on .gitlab/packaging/rpm.yml:
```diff
@@ -0,0 +1,49 @@
+  rules:
+    - !reference [.except_mergequeue]
+    - when: on_success
+    - when: on_success
```
```suggestion
```

chouquette committed:
remove duplicated rule

chouquette committed:
build a dedicated updater artifact for OCI packaging

we need 2 distinct builds for the "regular" packages and for the OCI
packages, at least for now

chouquette committed:
omnibus: updater: account for INSTALL_DIR provided by the env

chouquette committed:
ensure we can compute the OCI package version

chouquette commented:
/merge

raphaelgavache commented on .gitlab/packaging/rpm.yml:
```diff
@@ -0,0 +1,48 @@
+  before_script:
+    - source /root/.bashrc
+    - export RELEASE_VERSION=$RELEASE_VERSION_7
+    - export INSTALL_DIR=/opt/datadog-packages/datadog-agent/$(inv agent.version -u)-1
```
this looks like a left over as the path is /opt/datadog/updater for rpm

chouquette commented:
/merge --cancel

chouquette committed:
CI: add missing debian updater amd64 deploy job

chouquette committed:
fix deb/rpm packager INSTALL_DIR

chouquette committed:
fix deb x86_64 arch name

chouquette commented:
/merge
